### PR TITLE
feature(cuda): introduce cuda acceleration for the pbs and keyswitch

### PR DESCRIPTION
**Pull request by agnesLeroy**
### Description
- a new crate concrete-cuda is added to the repository, containing some
Cuda implementations for the bootstrap and keyswitch and a Rust wrapping
to call them
- a new backend_cuda is added to concrete-core, with dedicated entities
whose memory is located on the GPU and engines that call the Cuda
accelerated functions

### Checklist 

(Use '[x]' to check the checkboxes, or submit the PR and then click the checkboxes)

* [x] Tests for the changes have been added (for bug fixes / features)
* [x] Docs have been added / updated (for bug fixes / features)
* [x] The PR description links to the related issue (to link an issue, use '#XXX'.)
* [x] The tests on AWS have been launched and are successful (apply the `aws_test` to the PR to launch the tests on AWS)
* [x] The draft release description has been updated
* [x] Check for breaking changes and add them to commit message following the conventional commit [specification][conventional-breaking]

[conventional-breaking]: https://www.conventionalcommits.org/en/v1.0.0/#commit-message-with-description-and-breaking-change-footer

### CODE REVIEW

File: concrete-cuda/README.md [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+- `cuda_create_stream`, `cuda_destroy_stream`
+- `cuda_malloc`, `cuda_check_valid_malloc`
+- `cuda_memcpy_async_to_cpu`, `cuda_memcpy_async_to_gpu`
+- `cuda_get_number_of_gpus`
+- `cuda_synchronize_device`
+The only cryptographic operation it provides is an amortized implementation of the TFHE 
+programmable bootstrapping:
+- `cu_batch_bootstrap`
+
+These C++/CUDA functions are available to the [Concrete](https://github.com/zama-ai/concrete)
```
Comment (IceTDrinker) [inline concrete-cuda/README.md:19]: link is probably out of date ?

File: concrete-cuda/LICENSE [hunk 2]
```diff
@@ -0,0 +1,42 @@
+BSD 3-Clause Clear License
+
+Copyright © 2021 ZAMA.
```
Comment (IceTDrinker) [inline concrete-cuda/LICENSE:3]: License file is out of date, you should be able to re-use the one from the root of concrete-core

File: concrete-cuda/Cargo.toml [hunk 3]
```diff
@@ -0,0 +1,21 @@
+[package]
+name = "concrete-cuda"
+version = "0.1.0"
+edition = "2018"
```
Comment (IceTDrinker) [inline concrete-cuda/Cargo.toml:4]: 2021

File: concrete-cuda/Cargo.toml [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+authors = ["Joao Justi", "B. Barbakadze", "A. Leroy", "Zama team"]
+license = "BSD-3-Clause-Clear"
+description = "Cuda implementation for the concrete FHE library."
+homepage = "https://crates.io/crates/concrete"
```
Comment (IceTDrinker) [inline concrete-cuda/Cargo.toml:8]: homepage is not up to date, check other Cargo.toml for the link

File: concrete-cuda/Cargo.toml [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+documentation = "https://concrete.zama.ai"
+repository = "https://github.com/zama-ai/concrete"
```
Comment (IceTDrinker) [inline concrete-cuda/Cargo.toml:10]: repository is out of date

**Review by IceTDrinker — COMMENTED** [review]
Thanks @agnesLeroy for the PR!

As discussed I did a very high level review and did not get into the details of the Cuda or Rust code for concrete-cuda

File: concrete-tasks/src/check.rs [hunk 6]
```diff
@@ -12,7 +12,7 @@ lazy_static! {
 }
 
 pub fn doc() -> Result<(), Error> {
-    cmd!(<ENV_DOC_KATEX> &format!("cargo {} doc --features=doc --no-deps", get_nightly_toolchain()?))
+    cmd!(<ENV_DOC_KATEX> &format!("cargo {} doc --features=doc --no-deps --workspace --exclude concrete-cuda", get_nightly_toolchain()?))
```
[the remark this answers is not in the document]
Comment (agnesLeroy) [reply concrete-tasks/src/check.rs:15]: Not yet, I'll create it!

File: concrete-cuda/README.md (hunk 1 above)
Comment (agnesLeroy) [reply concrete-cuda/README.md:19]: Thanks for noticing!

File: concrete-cuda/LICENSE (hunk 2 above)
Comment (agnesLeroy) [reply concrete-cuda/LICENSE:3]: Oooh right, thanks a lot for catching this!

**Comment by agnesLeroy** [comment]
> Thanks @agnesLeroy for the PR!
> 
> As discussed I did a very high level review and did not get into the details of the Cuda or Rust code for concrete-cuda

That was actually super useful as many things had been done last year and were out of date for the links/license! Thank you very much!

File: concrete-tasks/src/test.rs [hunk 7]
```diff
@@ -37,13 +37,23 @@ pub fn ffi() -> Result<(), Error> {
...
 pub fn crates() -> Result<(), Error> {
     commons()?;
     core()?;
     core_test()?;
     csprng()?;
     npe()?;
-    ffi()
+    ffi()?;
+    cuda()?;
+    cuda_test()
```
[the remark this answers is not in the document]
Comment (agnesLeroy) [reply concrete-tasks/src/test.rs:56]: Oh no you're right I should remove them from here.

**Comment by IceTDrinker** [comment]
looks like the standard AWS tests still build or run the CUDA tests (the last failing build here: https://github.com/zama-ai/concrete-core/runs/7018469890?check_suite_focus=true)

**Comment by agnesLeroy** [comment]
> looks like the standard AWS tests still build or run the CUDA tests (the last failing build here: https://github.com/zama-ai/concrete-core/runs/7018469890?check_suite_focus=true)

Yes, my bad! I'll fix it :slightly_smiling_face:

**Comment by agnesLeroy** [comment]
@IceTDrinker only the test_uniformity has failed: can you approve so I can merge? Thx

**Comment by IceTDrinker** [comment]
Yep, as discussed ok since it was in an integration branch

**Review by IceTDrinker — APPROVED** [review]

**Pull request merged**